### PR TITLE
Support auto updates for Kubernetes workloads

### PR DESCRIPTION
**Pull request by vrothberg**
Add auto-update support to `podman kube play`.  Auto-update policies can    
be configured for:                                                          
 * the entire pod via the `io.containers.autoupdate` annotation             
 * a specific container via the `io.containers.autoupdate/$name` annotation 
                                                                            
To make use of rollbacks, the `io.containers.sdnotify` policy should be     
set to `container` such that the workload running _inside_ the container    
can send the READY message via the NOTIFY_SOCKET once ready.  For           
further details on auto updates and rollbacks, please refer to the          
specific article [1].                                                       
                                                                            
Since auto updates and rollbacks bases on Podman's systemd integration,     
the k8s YAML must be executed in the `podman-kube@` systemd template.       
For further details on how to run k8s YAML in systemd via Podman, please    
refer to the specific article [2].                                          
                                                                            
An examplary k8s YAML may look as follows:                                  
```YAML                                                                     
apiVersion: v1                                                              
kind: Pod                                                                   
metadata:                                                                   
  annotations:                                                              
      io.containers.autoupdate: "local"                                     
      io.containers.autoupdate/b: "registry"                                
  labels:                                                                   
    app: test                                                               
  name: test_pod                                                            
spec:                                                                       
  containers:                                                               
  - command:                                                                
    - top                                                                   
    image: alpine                                                           
    name: a                                                                 
  - command:                                                                
    - top                                                                   
    image: alpine                                                           
    name: b                                                                 
```                                                                         
                                                                            
[1] https://www.redhat.com/sysadmin/podman-auto-updates-rollbacks           
[2] https://www.redhat.com/sysadmin/kubernetes-workloads-podman-systemd     
                                                                            
Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>                     

#### Does this PR introduce a user-facing change?

```release-note
Support auto updates for `podman kube play`.
```

### CODE REVIEW

File: docs/source/markdown/podman-auto-update.1.md.in
Comment (richm) [inline docs/source/markdown/podman-auto-update.1.md.in:39]: sdnotify

Comment (vrothberg) [reply docs/source/markdown/podman-auto-update.1.md.in:39]: Thanks!

**Comment by vrothberg** [comment]
@edsantiago, I'd love to have your input on the tests.  They do not run fast for the known reasons.

**Comment by edsantiago** [comment]
Been looking. The refactoring is nice. Tests LGTM so far.

**Comment by vrothberg** [comment]
@containers/podman-maintainers PTAL

File: pkg/domain/infra/abi/play_utils.go
Comment (giuseppe) [inline pkg/domain/infra/abi/play_utils.go:28]: is `getAutoUpdateAnnotation` only used by tests?  Would it make sense to move it to the `_test.go` file?

**Review by giuseppe — APPROVED** [review]
one nit, otherwise LGTM

File: pkg/domain/infra/abi/play_utils.go
Comment (vrothberg) [reply pkg/domain/infra/abi/play_utils.go:28]: Nice catch.  I forget to plug it in and it turns out the backend already handles the auto-update label.  I'll drop the commit.

**Comment by vrothberg** [comment]
@rhatdan PTAL

**Comment by vrothberg** [comment]
@edsantiago there was a flake in the new test on ARM.  I will keep an eye on it.  Please let me know if you see it flake another time.

**Comment by edsantiago** [comment]
You mean [this one?](https://api.cirrus-ci.com/v1/artifact/task/6390523807137792/html/sys-podman-fedora-36-aarch64-root-host.log.html#t--00285)
```
         # #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
         # #|     FAIL: container-specified auto-update policy gets applied
         # #| expected: '.*podman-kube@-tmp-podman_bats.GNDJSD-test.yaml.service,.* (test_pod-b),quay.io/libpod/testimage:20220615,false,registry.*' (using expr)
         # #|   actual: 'podman-kube@-tmp-podman_bats.GNDJSD-test.yaml.service,4107fa87fd9e (test_pod-b),quay.io/libpod/testimage:20220615,false,local'
         # #|         > 'podman-kube@-tmp-podman_bats.GNDJSD-test.yaml.service,e7cd7e8a0a90 (test_pod-a),quay.io/libpod/testimage:20220615,false,local'
         # #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
It looks like the `registry` string is missing. That is not likely to be a one-off. Is there any way you can look into it before this merges?

**Comment by vrothberg** [comment]
> It looks like the registry string is missing. That is not likely to be a one-off. Is there any way you can look into it before this merges?

One container seems to have the wrong policy set.  Unfortunately, I do not manage to reproduce the issue at all.

**Review by edsantiago — CHANGES_REQUESTED** [review]
Lots of [red](https://api.cirrus-ci.com/v1/artifact/task/5827573853716480/html/sys-podman-fedora-36-root-host.log.html#t--00296); needs cleanup.

Also, I just got a different flake when testing locally:
```
  # .../bin/podman auto-update --dry-run --format {{.Unit}},{{.Container}},{{.Image}},{{.Updated}},{{.Policy}}
   podman-kube@-tmp-podman_bats.mSkYdc-test.yaml.service,1549ec899b55 (test_pod-b),podman.io/image:test,failed,registry
   podman-kube@-tmp-podman_bats.mSkYdc-test.yaml.service,af82e5c104d2 (test_pod-a),quay.io/libpod/testimage:20220615,false,registry
   Error: checking image updates for container 1549ec899b5599cab58d3d09cb198e7db3236095a963d8e8339e9552afb41386: pinging container registry podman.io: invalid status code from registry 404 (Not Found)
   [ rc=125 (** EXPECTED 0 **) ]
```

Sorry, gotta block.

**Comment by edsantiago** [comment]
The red cleanup warnings can be fixed with
```diff
index ea656e6f6..85560e2ef 100644
--- a/test/system/255-auto-update.bats
+++ b/test/system/255-auto-update.bats
@@ -504,7 +504,7 @@ EOF                                                                                                                                        
 
     # Clean up
     systemctl stop $service_name
-    run_podman rmi $(pause_image)
+    run_podman rmi $local_image $(pause_image)
     rm -f $UNIT_DIR/$unit_name
 }
 
```
And, I'm reliably seeing the `podman.io 404` error on my laptop (as root, not as rootless).

**Comment by edsantiago** [comment]
Something is nondeterministically pinging the registry. I think `--dry-run` is broken. This may or may not be the fault of this PR, but it has to be fixed before this PR merges. Again, sorry.

Confirmation: this change:
```diff
-    local_image=podman.io/image:test
+    local_image=localhost/localimage:$(random_string 10)
```
changes the flake error to:
```
Error: checking image updates for container [snip]: pinging container registry localhost: Get "https://localhost/v2/": dial tcp [::1]:443: connect: connection refused
```
(i.e., from `podman.io/404` to `localhost/ECONNREFUSED`)

**Comment by vrothberg** [comment]
Brilliant work, @edsantiago!  Thanks a bunch!

**Comment by vrothberg** [comment]
@edsantiago PTanotherL. The issue was as follows. The `registry` policy instructs auto updates to look at the registry and check whether there's a new image.  Errors when looking up a new image were treated as fatal which I think is wrong - especially in an Edge use case.  The error is now non-fatal but logged.

I also massaged the test a bit and ironed out some farts in the code.

File: test/system/255-auto-update.bats
Comment (edsantiago) [inline test/system/255-auto-update.bats:507]: The last two need to be curlies, not parens. Parens mean (execute this as a command); curlies means "variable". You may also want to add oldID:
```bash
    run_podman rmi -f $(pause_image) $local_image $newID $oldID
```
(I removed the curlies altogether because I find it less confusing this way, no mixing of parens. But, up to you.)

Comment (vrothberg) [reply test/system/255-auto-update.bats:507]: Thanks!  The values of {new,old}ID were bogus as well as I ran `podman inspect` instead of `podman_run inspect`

File: test/system/255-auto-update.bats
Comment (edsantiago) [inline test/system/255-auto-update.bats:487]: OOPS! Needs to be `run_podman`, not podman!

**Review by edsantiago — COMMENTED** [review]
also, longstanding bug: `_confirm_update()` is missing a `timeout=$((timeout - 1))`

File: test/system/255-auto-update.bats
Comment (edsantiago) [inline test/system/255-auto-update.bats:493]: Is there a way to override the timeout here? 90s is a long time in test-land.

Comment (vrothberg) [reply test/system/255-auto-update.bats:493]: Can you elaborate which timeout you are referring to?  I don´t see one for `auto-update`.

**Comment by vrothberg** [comment]
Looks like I need to re-add parts of the code I removed.  Will take care of that on Monday :^)

File: test/system/255-auto-update.bats
Comment (edsantiago) [reply test/system/255-auto-update.bats:493]: This command (line 493) takes 90 seconds to run. Confirm via:
```diff
+ echo "# $(date) got here 1" >&3
  run_podman auto-update ...
+ echo "# $(date) got here 2" >&3
```
Is there a way to make that command take a shorter time?

File: docs/source/markdown/podman-auto-update.1.md.in
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-auto-update.1.md.in:34]:
```suggestion
Podman supports auto-updates for Kubernetes workloads.  As mentioned above, `podman auto-update` requires the containers to be running systemd.  Podman ships with a systemd template that can be instantiated with a YAML file on the systems unit file, see podman-generate-systemd(1).
```

File: test/system/255-auto-update.bats
Comment (vrothberg) [reply test/system/255-auto-update.bats:493]: Got it now.  Again, thanks a ton for finding that.

There really was a deadlock/infinite wait when a container dies without sending the READY message.  The issue has now been resolved and I'll push an updated version in a jiffy.

File: pkg/systemd/notifyproxy/notifyproxy.go
Comment (Luap99) [inline pkg/systemd/notifyproxy/notifyproxy.go:130]: This check seems fragile since we use a fixed buffer, what happens if the ready message starts at byte 1022? In this case the string will not match and the next read will also not match since the string was cut in two pieces.
Is there a guarantee that sd notify messages are always less then 1024 byte? Otherwise you would need to read until a newline or EOF.

File: docs/source/markdown/podman-auto-update.1.md.in
Comment (Luap99) [inline docs/source/markdown/podman-auto-update.1.md.in:40]: What is the default?

File: pkg/systemd/notifyproxy/notifyproxy.go
Comment (vrothberg) [reply pkg/systemd/notifyproxy/notifyproxy.go:130]: I think it doesn't hurt to make it more robust.  I found it to be sufficient since only the READY message is supported.

**Comment by vrothberg** [comment]
I think we're good to go

**Review by edsantiago — APPROVED** [review]

**Pull request merged**
